ambry-commons/src/main/java/com.github.ambry.commons/ByteBufferReadableStreamChannel.java

```diff
@@ -65,6 +74,30 @@ public long getSize() {
   }
 
   @Override
+  public void setDigestAlgorithm(String digestAlgorithm)
+      throws NoSuchAlgorithmException {
+    if (digest != null && digest.getAlgorithm().equals(digestAlgorithm)) {
+      return;
+    }
+    digest = MessageDigest.getInstance(digestAlgorithm);
+    digestBytes = null;
+  }
+
+  @Override
+  public byte[] getDigest() {
+    if (digest == null) {
+      return null;
+    } else if (digestBytes == null) {
+      int savedPosition = buffer.position();
+      buffer.position(startPos);
+      digest.update(buffer);
+      buffer.position(savedPosition);
+      digestBytes = digest.digest();
```

Comment:
if you use `ByteBuffer#slice()` in the constructor, I think you wouldn't have to save `size` and `startPos`. This piece of code would then be (I think):
```java
buffer.mark();
buffer.rewind();
digest.update(buffer);
buffer.reset();
```
`slice()` creates a new object though just like `wrap()`, and this does not seem like much of an improvement either - so weigh that in as well.